libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/NavigationRouteEx.kt

```diff
@@ -0,0 +1,170 @@
+@file:JvmName("NavigationRouteUtils")
+
+package com.mapbox.navigation.base.utils.route
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.RouteLeg
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.utils.DecodeUtils.stepGeometryToPoints
+import com.mapbox.navigation.utils.internal.logD
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+private const val LOG_CATEGORY = "NavigationRouteUtils"
+
+/**
+ * This function checks whether the [NavigationRoute] has unexpected closures, which could be a reason to re-route.
+ *
+ * `true` is returned whenever there's any [RouteLeg.closures] outside of a direct region around a coordinate from [RouteOptions.coordinatesList]
+ * that was allowed for snapping to closures with [RouteOptions.snappingIncludeClosuresList] or [RouteOptions.snappingIncludeStaticClosuresList].
+ * flags. Otherwise, `false` is returned.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+suspend fun NavigationRoute.hasUnexpectedClosures(): Boolean =
+    withContext(Dispatchers.Default) {
+        val snappingResultList = directionsRoute.getSnappingResultList()
+
+        var currentLegFirstWaypointIndex = 0
+        directionsRoute.legs()?.forEachIndexed { routeLegIndex, routeLeg ->
+
+            val legLastGeometryIndex by lazy {
+                directionsRoute.stepsGeometryToPoints(routeLeg).lastIndex
+            }
+
+            val silentWaypoints = routeLeg.silentWaypoints()
+
+            routeLeg.closures()?.forEach { closure ->
+                val silentWaypointsInClosureRange = silentWaypoints.inGeometryRange(
+                    closure.geometryIndexStart(), closure.geometryIndexEnd()
+                )
+                if (silentWaypointsInClosureRange.isNotEmpty()) {
```

Comment:
Why do you need this check? Can't you just rely on forEach not being executed for an empty collection?

Reply:
yes, it can. Non needed check 👍